cpp/src/arrow/dataset/discovery.cc

```diff
@@ -134,8 +135,26 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
 Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
     std::shared_ptr<fs::FileSystem> filesystem, const std::vector<fs::FileInfo>& files,
     std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options) {
+  // Discover files in directories and globs
+  std::vector<fs::FileInfo> discovered_files;
+  for (const auto& file : files) {
+    if (file.IsDirectory()) {
+      fs::FileSelector file_selector;
+      file_selector.base_dir = file.dir_name();
+      file_selector.recursive = true;
+      ARROW_ASSIGN_OR_RAISE(auto folder_files, filesystem->GetFileInfo(file_selector));
+      std::move(folder_files.begin(), folder_files.end(),
+                std::back_inserter(discovered_files));
+    } else if (file.IsGlob()) {
+      ARROW_ASSIGN_OR_RAISE(auto files, filesystem->GetFileInfoGlob(file));
+      std::move(files.begin(), files.end(), std::back_inserter(discovered_files));
+    } else if (file.IsFile()) {
+      discovered_files.emplace_back(file);
+    }
+  }
+
```

Comment:
So this is something of an abstract concern but I think we have a bit of ambiguity between input & output types.

I see `FileInfo` as an "output type".  It is something you get back from a filesystem that describes something that it went out and discovered or inspected.

In the meantime, PlatformFilename and FileSelector are "input" types.  They are provided by the user for the purpose of obtaining FileInfo objects.

As a result I'm not sure it makes sense to have `FileType::Glob` because I would interpret that as the filesystem discovering a glob inside the filesystem.  For example `GetFileInfo` would never return `FileType::Glob`.

Instead, can we do something slightly different:

Instead of modifying the `Make` overload that accepts `const std::vector<fs::FileInfo>&` can we add a new overload that takes `const std::vector<fs::FileSelector>&`?

Then, instead of adding `FileType::Glob` can we add an `is_glob` property to `FileSelector` (if this is true then `base_dir` is expected to be a glob pattern and `recursive` will be ignored).

We could also then change the rules a little on `FileSelector` so that `base_dir` is allowed to represent a single file.  Then we can treat the "path" case (user doesn't specify it as a file or a directory) as a selector also.  That would allow us to move all the `GetFileInfo` calls out of the Substrait consumer.

CC @pitrou for a second opinion as he has a lot of experience here.

Reply:
File selector might be able to handle files, directories, and globs, isn't it? So, we'll have an extra property like is_glob for glob discovery. But, can we have a property that tell us if it is a file, a directory, or a glob? Just to make it more user-friendly and avoid having a property per each type.

Reply:
I'm not entirely sure what you are describing but I think it sounds ok.  Can you show an example of what the fields would look like?  For example, I was thinking:

```
struct ARROW_EXPORT FileSelector {
  std::string base_dir;
  bool allow_not_found;
  bool recursive;
  bool is_glob;
  int32_t max_recursion;
};
```

Are you thinking of using an enum instead of having both `bool is_glob` and `bool recursive`?  We could have an `is_file` property but I don't know that we would use it.  If we know we have a file we can just use a filename.  If we have a path then we don't know if it is a file or not.

Reply:
Yes, I'm thinking of using an enum, something like this:

```
struct ARROW_EXPORT FileSelector {
  std::string base_dir;
  FileType type;
  bool allow_not_found;
  int32_t max_recursion;
  
  FileSelector() : type(Unknown), allow_not_found(false), recursive(false), 
                   max_recursion(INT32_MAX) {}
  FileSelector(std::string base_dir) : base_dir(base_dir), type(File), 
                                       allow_not_found(false), recursive(false), 
                                       max_recursion(INT32_MAX) {}
};
```

Reply:
Ok, yes, I think an enum approach is fine.  I would maybe call it `SelectorType` instead of `FileType` since we already have something named `FileType`.  I would have a slight preference to not reuse `FileType` because that will lead to some confusion (users creating a selector will wonder what "unknown" means and users getting fileinfo will wonder what "glob" means).

Also, we will want to make sure it is clear (in comments at least to start but later in documentation) the meaning of the fields.

Reply:
Ok, I understand. So, should I go ahead and perform the changes? Or should we be waiting for someone else opinion?

Reply:
If you have time to perform the changes I think it's good to go and make them.

Reply:
Ok, I'm working on that. 

@westonpace, I have a question about current use of `FileSelector`. Some tests use `selector.recursive = true; `, I think recursive in the new implementation means `selector.type = Directory` and `selector.max_recursion = INT32_MAX`. So, we should add a method like `set_recursive(bool)` and `bool is_recursive()` or setting `selector.max_recursion` is enough?